CI: Add Kafka specific tests to make sure kafka cluster is up correctly.

manalibhutiyani opened the pull request:
These changes add kafka specific tests to ensure that the broker is up correctly before trying to create topics/produce/consume to it.

It also increases the session and connection timeout from 20 sec from 60 sec, due to logs observed
https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-K8s/350/ 
which cause the kafka broker to timeout since zookeeper has not responded for 20 sec.


Fixes: #4431
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

```release-note
Add Kafka specific CI test checks to make sure kafka cluster is up correctly.
```

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/4488)
<!-- Reviewable:end -->

Commit:
test/manifest: Remove automatic topic creation on container start

This commit removes the automatic topic creation from `wurstmeister/kafka-docker` image,
because `wurstmeister/kafka-docker` does not guarentee that the Kafka cluster is up and running, and zookeeper
is aware of all the kafka brokers before creating these topics.
This is an attempt to make sure that we wait for enough time and make sure that
the kafka broker/kafka-zookeeper cluster is ready before creating any topics.

Fixes: #4431
Signed-Off-By: Manali Bhutiyani

Commit:
test/runtime:  Remove hardcoded timeouts in the kafka runtime test

This commit removes the hardcoded timeouts and uses the `helpers.HelperTimeout` instead.

Fixes: #4431
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

Commit:
test/runtime: Change check from listTopics to createTopics to check if Kafka broker is up.

This change fixes the check in `waitForKafkaBroker` from `listTopicsCmd` to `createTopicCmd`.
It was done because, the `kafka-topics.sh --list` scripts query the zookeeper directly for topics, and even if
the kafka broker is not up, list topics returns success. This was an incorrect way to check if the
kafka broker is up or not. This change tries to create a test-topic and waits until the topic is succesfully created
or times out. This will ensure that the kafka broker has come up correctly before trying to produce/consume to it.

Fixes: #4431
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

Commit:
test/K8s: Wait for kafka broker to be up correctly.

The k8s livenessProbe/readinessProbe only guarentees that the pod is up and the port is open. This does not necessarily
mean that the kafka broker is up or that zookeeper is aware of all the kafka brokers in the cluster and has done the necessary setup.
This change tries to create a test-topic and waits till the topic is succesfully created
or times outs. This will ensure that the kafka broker has come up correctly before trying to produce/consume to it.

Fixes: #4431
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

rlenglet commented on test/k8sT/KafkaPolicies.go:
```diff
@@ -57,9 +60,30 @@ var _ = Describe("K8sValidatedKafkaPolicyTest", func() {
+		}
+
+		// WaitKafkaBroker waits for the broker to be ready, by executing
+		// a produce request on existing topics and waiting for a response from broker
```
`a produce request on existing topics and waiting for a response from broker` -> `a command repeatedly until it succeeds, or a timeout occurs`

manalibhutiyani replied:
done. Also changed wording in `runtime` tests

rlenglet reviewed (approved):
LGTM. Just one nitpick re: a comment.

ianvernon commented on test/runtime/kafka.go:
```diff
@@ -34,13 +34,10 @@ var _ = Describe("RuntimeValidatedKafka", func() {
 
```
Typo in commit: "This was an incorrect way to check if the kafka broker if up or not" --> "This was an incorrect way to check if the kafka broker is up or not

ianvernon replied:
"waits till" --> "waits until"
"times outs" --> "times out".

manalibhutiyani replied:
done

ianvernon commented on test/runtime/kafka.go:
```diff
@@ -190,14 +187,8 @@ var _ = Describe("RuntimeValidatedKafka", func() {
-
 		By("Sending produce request on kafka topic `allowedTopic`")
-		for i := 1; i <= MaxMessages-1; i++ {
+		for i := 1; i <= MaxMessages; i++ {
```
Can you explain the change for this conditional check?

manalibhutiyani replied:
Earlier we were using the first produce request as a way to check the broker as well. Hence we only produced 1 to  MaxMessages - 1. Now this check was changed to test `createTopics` instead and hence we are producing all messages, i.e MaxMessages

ianvernon reviewed (dismissed):
Small nits.

Commit:
test/k8s: Increase kafka-zookeeper session/connection timeout to 60 sec

It was observed in one of the CI test failures, that zookeeper can take more than
20ms (current configured session timeout) to respond to kafka broker causing
kafka broker to crash loop. This commit increases the session as well as timeout in the CI setup

Fixes: #4431
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

Commit:
test/runtime: Increase kafka-zookeeper session/connection timeout to 60 sec

It was observed in one of the CI test failures, that zookeeper can take more than
20ms (current configured session timeout) to respond to kafka broker causing
kafka broker to crash loop. This commit increases the session as well as timeout in the CI setup

Fixes: #4431
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>

Commit:
docs/examples: Increase kafka-zookeeper session/connection timeout to 60 sec

It was observed in one of the CI test failures, that zookeeper can take more than
20ms (current configured session timeout) to respond to kafka broker causing
kafka broker to crash loop. This commit increases the timeout in the GSG as well
to reflect the new timeout in CI.

Fixes: #4431
Signed-Off-By: Manali Bhutiyani <manali@covalent.io>
